src/mqtt_interface.rs

```diff
@@ -0,0 +1,259 @@
+use super::StringSet;
+use core::fmt::Write;
+use heapless::{consts, String};
+use minimq::{
+    embedded_nal::{IpAddr, TcpStack},
+    MqttClient, Property, QoS,
+};
+
+#[derive(Debug)]
+pub enum Error<E: core::fmt::Debug> {
+    /// The provided device ID is too long.
+    IdTooLong,
+
+    /// MQTT encountered an internal error.
+    Mqtt(minimq::Error<E>),
+
+    /// The network stack encountered an error.
+    Network(E),
+}
+
+impl<E: core::fmt::Debug> From<minimq::Error<E>> for Error<E> {
+    fn from(err: minimq::Error<E>) -> Self {
+        match err {
+            minimq::Error::Network(err) => Error::Network(err),
+            other => Error::Mqtt(other),
+        }
+    }
+}
+
+/// An action that applications should act upon.
+#[derive(PartialEq)]
+pub enum Action {
+    /// There is nothing to do except continue normal execution.
+    Continue,
+
+    /// The settings are being commit to memory. The application should take steps to make staged
+    /// settings active.
+    CommitSettings,
+}
+
+// Generate an MQTT topic of the form `<device_id>/<topic>`.
+//
+// # Returns
+// The string - otherwise, an error indicating the generated string was too long.
+fn generate_topic<'a, 'b>(device_id: &'a str, topic: &'b str) -> Result<String<consts::U128>, ()> {
+    let mut string: String<consts::U128> = String::new();
+    write!(&mut string, "{}/{}", device_id, topic).or(Err(()))?;
+    Ok(string)
+}
+
+/// An interface for managing MQTT settings.
+pub struct MqttInterface<T, S>
+where
+    T: StringSet,
+    S: TcpStack,
+{
+    // TODO: Allow the user to specify buffer size.
+    client: Option<MqttClient<minimq::consts::U256, S>>,
+    pub settings: T,
+
+    subscribed: bool,
+    settings_topic: String<consts::U128>,
+    commit_topic: String<consts::U128>,
+    default_response_topic: String<consts::U128>,
+    id: String<consts::U128>,
+}
+
+impl<T, S> MqttInterface<T, S>
+where
+    T: StringSet,
+    S: TcpStack,
+{
+    /// Construct a new settings interface using the network stack.
+    ///
+    /// # Args
+    /// * `stack` - The TCP network stack to use for communication.
+    /// * `id` - The ID for uniquely identifying the device. This must conform with MQTT client-id
+    ///          rules. Specifically, only alpha-numeric values are allowed.
+    /// * `broker` - The IpAddress of the MQTT broker.
+    /// * `settings` - The initial settings of the interface.
+    ///
+    /// # Returns
+    /// A new `MqttInterface` object that can be used for settings configuration and telemtry.
+    pub fn new(
+        stack: S,
+        id: &str,
```

Comment:
We're treating the client ID and the prefix ID identically here, but there are different rules that MQTT enforces for these values. Perhaps we need to allow a client-id and a prefix.

It may make more sense to allow the user to supply an MqttClient, as this would allow them to specify an arbitrary ID, broker, etc. Then, the prefix of the device could be selectable.